usvm-core/src/main/kotlin/org/usvm/statistics/UDebugProfileObserver.kt

```diff
@@ -0,0 +1,289 @@
+package org.usvm.statistics
+
+import mu.KLogging
+import org.usvm.PathNode
+import org.usvm.StateId
+import org.usvm.UState
+import org.usvm.algorithms.TrieNode
+import java.util.concurrent.atomic.LongAdder
+
+private typealias StatsCounter = LongAdder
+
+private val logger = object : KLogging() {}.logger
+
+/**
+ * Collect UMachine profile for debug purposes.
+ * */
+open class UDebugProfileObserver<Statement, Method, State : UState<*, Method, Statement, *, *, *>>(
+    private val statementOperations: StatementOperations<Statement, Method>,
+    private val profilerOptions: Options<Statement, State> = Options(),
+) : UMachineObserver<State> {
+
+    interface StatementOperations<Statement, Method> {
+
+        // mandatory operations
+
+        fun getMethodOfStatement(statement: Statement): Method
+
+        fun getStatementIndexInMethod(statement: Statement): Int
+
+        /**
+         * Return null if statement is not a call statement
+         * */
+        fun getMethodToCallIfCallStatement(statement: Statement): Method?
+
+        // operations with default implementations
+
+        fun printStatement(statement: Statement): String = statement.toString()
+        fun printMethodName(method: Method): String = method.toString()
+        fun getOriginalInst(statement: Statement): Statement = statement
```

Comment:
This is a jvm specific hack. So, let's add a comment about this to avoid confusion about what the original inst is.